### ISSUE TITLE
Syntax and font-lock for Racket here strings

### ISSUE THREAD

**Issue by samth** [greghendershott/racket-mode#176]
See this file: https://github.com/jbclements/morse-code-trainer/blob/71b8d2007ad741afac995d6b065ed9b6dc3af416/morse-code-trainer/morse-code-table.rkt

**Comment by samth** [comment]
The problem is syntax coloring of here-strings.

**Comment by greghendershott** [comment]
I see what you mean about here strings.

Just to make sure I'm not missing something else, when you mentioned unicode, is that actually N/A?

As for here-strings. I can improve this. I'm a little cautious because multi-line font-lock can be tricky to make both (a) perfect and (b) fast. Identification of things like multi-line sexp comments is helped by Emacs' ability to parse sexps quickly. Searching back to point-min for `#<<` would obviously suck. It would help if I can limit the search for a `#<<` to `beginning-of-defun`. Doing so would mean that a here-string sitting at module/top level -- not even inside a `(define )` form -- could be missed. But (a) who would even do that and (b) even that would often be font-locked correctly, for instance when visiting the file and navigating forward.

**Comment by samth** [comment]
Well, it might be a combination of both unicode and here-strings, but it sounds like you figured out that it isn't.

As for speed, is font-lock not run from the top of the buffer?

**Comment by greghendershott** [comment]
> As for speed, is font-lock not run from the top of the buffer?

Not necessarily. That would for example make edits near the end of a long file quite slow. By default it's more conservative about the region to be re-fontified. A major mode can extend the region to cover multi-line constructs it knows about, but you don't want that determination itself to be slow and/or the answer to be unnecessarily large and therefore slow.

That's for font-lock. There's a similar issue and process for determining the char-syntax class of characters in the buffer (i.e. is this an opening paren, a comment, a string, etc.).  And actually I think this should be a char-syntax thing. A here-string is a _string_.  Probably the syntax will drive the font-lock as it does for ordinary strings. I think. I'll dig in and figure out the best way.

**Comment by greghendershott** [comment]
> > Just to make sure I'm not missing something else, when you mentioned unicode, is that actually N/A?

> Well, it might be a combination of both unicode and here-strings, but it sounds like you figured out that it isn't.

It's just that I don't see what you mean by "unicode" here.  So checking if there's something you see/know that I'm failing to.

---

As for here strings: I've spent quite some time working on this, and then also peeked at how `sh-mode` handles here docs. I have something working in general.

Unfortunately it can be horrifically slow with that example file -- however, it's equally slow if you replace the here string `#<<ID\n` and `\nID\n` delimiters with plain old `"` character delimiters. So, there's something else going on, that I need to understand. It may be related to all the `{` `}` and `'` characters in the multi-line string, and using `syntax-ppss`, but I'm not sure yet.

TL;DR Although I believe I've already implemented here strings just fine, in the process I managed to peel the onion and discover a possibly unrelated performance problem. So I'm detouring to figure that out, too.

**Comment by samth** [comment]
As to Unicode, all I meant was that the string is not ASCII-only, which I imagined could potentially be an issue.

**Comment by greghendershott** [comment]
Oh, now I finally realize those are composition and en-dash chars, `·` and `–`. In the wall of text I thought they were `.` and `-`. Didn't try to understand what the Racket code _does_; I was looking at it like a lexer. :)

---

Update: I figured out the performance problem, and fine-tuned some related things. I want to test more.

Also consider what sort of commit history would more clearly show the various changes, then create and push that.
